blockchain/chain.go

```diff
@@ -1290,6 +1290,176 @@ func (b *BlockChain) HeightRange(startHeight, endHeight int32) ([]chainhash.Hash
 	return hashes, nil
 }
 
+// locateBlocks returns the hashes of the blocks after the first known block in
+// the locator until the provided stop hash is reached, or up to the provided
+// max number of block hashes.
+//
+// See the comment on the exported function for more details on special cases.
+//
+// This function MUST be called with the chain state lock held (for reads).
+func (b *BlockChain) locateBlocks(locator BlockLocator, hashStop *chainhash.Hash, maxHashes uint32) []chainhash.Hash {
+	// There are no block locators so a specific block is being requested
+	// as identified by the stop hash.
+	stopNode := b.index.LookupNode(hashStop)
+	if len(locator) == 0 {
+		// No blocks with the stop hash were found so there is nothing
+		// to do.
+		if stopNode == nil {
+			return nil
+		}
+		return []chainhash.Hash{*hashStop}
+	}
+
+	// Find the most recent locator block hash in the main chain.  In the
+	// case none of the hashes in the locator are in the main chain, fall
+	// back to the genesis block.
+	knownNode := b.bestChain.Genesis()
+	for _, hash := range locator {
+		node := b.index.LookupNode(hash)
+		if node != nil && b.bestChain.Contains(node) {
+			knownNode = node
+			break
+		}
+	}
+
+	// Start at the block after the most most recently known block.  When
+	// there is no next block it means the most recently known block is
+	// the tip of the best chain, so there is nothing more to do.
+	knownNode = b.bestChain.Next(knownNode)
```

Comment:
Great suggestion.  I was thinking of it from the point of view of the local peer, but I I agree your name is much clearer.  Thanks!  Changed.